From 4f19eb1db6d2e2adb33d19bcb3fce420ad8c4dad Mon Sep 17 00:00:00 2001
From: Christine Gharzuzi <cgharzuzi@habana.ai>
Date: Thu, 13 Feb 2020 19:18:47 +0200
Subject: [PATCH] [SW-0] goya2: spi dma copy WA

Due to HW issue with spi DMA copy, an extra dummy copy
with a different size than the intended is needed,
otherwise spi DMA stops operating

Change-Id: I4b453cea677bdb9059b71bb002a0e92ab7f7da76
Signed-off-by: Christine Gharzuzi <cgharzuzi@habana.ai>
---

diff --git a/common/dma.c b/common/dma.c
index 62311f9..63532d5 100644
--- a/common/dma.c
+++ b/common/dma.c
@@ -9,8 +9,8 @@
 #include "dma.h"
 #include "spi_flash.h"
 
-void dma_copy_image(uint16_t image_size, uint32_t flash_offset,
-		    uint64_t dest_offset, uint8_t fw_sram_dst)
+static void dma_copy_image_internal(uint16_t image_size, uint32_t flash_offset,
+		uint64_t dest_offset, uint8_t fw_sram_dst)
 {
 	uint32_t reg_spi_dma_ctrl_s =  0;
 	uint32_t spi_dma_dst_addr_l = DMA_ADDRESS_LOW(dest_offset);
@@ -40,6 +40,15 @@
 	regmap_writel(reg_spi_dma_ctrl_s, mmPSOC_GLOBAL_CONF_SPI_DMA_CTRL);
 }
 
+void dma_copy_image(uint16_t image_size, uint32_t flash_offset,
+		    uint64_t dest_offset, uint8_t fw_sram_dst)
+{
+	dma_copy_image_internal(4, flash_offset, dest_offset, fw_sram_dst);
+	dma_wait_finish();
+	dma_copy_image_internal(image_size, flash_offset, dest_offset,
+			fw_sram_dst);
+}
+
 int dma_wait_finish(void)
 {
 	uint32_t reg_spi_dma_status_s = 0;
@@ -54,4 +63,4 @@
 		     PSOC_GLOBAL_CONF_SPI_DMA_STATUS_ERROR_SHIFT;
 
 	return crc_error;
-}
\ No newline at end of file
+}
